### PR TITLE
reader: add a reader interface

### PR DESCRIPTION
**Pull request by scolsen**
This commit makes read a generic interface and provides two
implementations, byte reader and string reader. The string reader is
just the original read function. The byte reader reads files byte by
byte.

This interface allows users to define read implementations for arbitrary
data types.

### CODE REVIEW

**Review by hellerve — COMMENTED** [review]
Great idea!

One thing that comes to mind: since the error handling on non-readability should always remain the same, maybe we could factor that out? Something like this:

    ; your read interface
    (definterface read-contents (Fn [(Ref File) Int] (Result a String)))
    
    (defmodule File
        (doc read "reads `len` characters from a file. Generic on its return value, custom readers can be added by implementing `read-contents`.
        
     Example:
     ```
     (IO.println &(read f 10)) ; => "Hello, Joe" (assuming that’s the file contents)
     (String.length &(read f 10)) ; => 10
     
     (IO.println &(Array.str &(read f 10))) ; => [\H \e \l \l \o \, \ \J \o \e]
     (Array.length &(read f 10)) ; => 10
     ```")
        (defn read [f len]
          (if (readable? f)
            (read f len)
            (Result.Error (fmt "The file “%s” is not readable" (name f)))))

**Comment by scolsen** [comment]
> Great idea!
> 
> One thing that comes to mind: since the error handling on non-readability should always remain the same, maybe we could factor that out? Something like this:
> 
> ```
> ; your read interface
> (definterface read-contents (Fn [(Ref File) Int] (Result a String)))
> 
> (defmodule File
>     (doc read "reads `len` characters from a file. Generic on its return value, custom readers can be added by implementing `read-contents`.
>     
>  Example:
>  ```
>  (IO.println &(read f 10)) ; => "Hello, Joe" (assuming that’s the file contents)
>  (String.length &(read f 10)) ; => 10
>  
>  (IO.println &(Array.str &(read f 10))) ; => [\H \e \l \l \o \, \ \J \o \e]
>  (Array.length &(read f 10)) ; => 10
>  ```")
>     (defn read [f len]
>       (if (readable? f)
>         (read f len)
>         (Result.Error (fmt "The file “%s” is not readable" (name f)))))
> ```

Good call! I'll add that in.

**Comment by scolsen** [comment]
added!

File: file.carp [hunk 1]
```diff
@@ -258,6 +258,27 @@ writable.")
...
+    ("This function is generic in its return argument. You can " false)
+    ("provide a custom reader implementation by implementing the " false)
+    "read-from-file interface."
+    ""
+    ("The surrounding context typically determine's the return type." false)
+    ("if no implementation of read-from-file exists for the type, you'll " false)
+    "need to provide one. You can also use `the` to specify the return type."
+    "For example (to read the first 3 bytes of the file):"
+    "```"
+    "(match (File.open-with \"test-file.txt\" \"r\")"
```
Comment (hellerve) [inline file.carp:273]: To avoid having to write it twice, you can put the `the` in the first clause of `match`

File: src/byte-reader.carp [hunk 2]
```diff
@@ -2,20 +2,19 @@
 
 (defmodule File
   (defmodule ByteReader
+    ;; TODO: Use the safer IO.fgetc once in HEAD
```
Comment (hellerve) [inline src/byte-reader.carp]: Should we wait for this to be merged?

File: src/byte-reader.carp [hunk 3]
```diff
@@ -2,20 +2,19 @@
...
-      (if (readable? f)
-          (let-do [bytes (Array.allocate n)
-                   i 0]
-            (while-do (< i n)
-              (Array.aset-uninitialized! &bytes i (read-byte @(file f)))
-              (++ i))
-            (Result.Success bytes))
-          (Result.Error (fmt "The file %s is not readable" (name f)))))
-    (implements read read)
+    (doc read-bytes "Read n bytes from a file.")
```
Comment (hellerve) [inline src/byte-reader.carp]:
```suggestion
    (doc read-bytes "Read `n` bytes from a file.")
```

File: src/byte-reader.carp (hunk 2 above)
Comment (hellerve) [reply src/byte-reader.carp]: Looks like it just got merged :)

Comment (scolsen) [reply src/byte-reader.carp]: Done!

File: file.carp (hunk 1 above)
Comment (scolsen) [reply file.carp:273]: Thanks!

**Review by hellerve — APPROVED** [review]
Thank you 🚀

**Pull request merged**